File: pkg/earthsci-ast-go/pkg/esm/serialize.go
package esm

import (
	"encoding/json"
	"fmt"
	"os"
)

// marshalCanonical pre-processes v with canonicalizeForJSON so every float
// is emitted in discretization RFC §5.4.6 form (trailing ".0" for
// integer-valued magnitudes in [−1e21+1, 1e21−1], exponent form outside
// that range) before running encoding/json. Without this pass Go emits
// float64(1.0) as "1", which collides with int64(1) on the wire and
// breaks the round-trip int/float node distinction.
func marshalCanonical(v any, indent bool) ([]byte, error) {
	canonical, err := canonicalizeForJSON(v)
	if err != nil {
		return nil, err
	}
	canonical, err = restoreUnresolvedTopLevelMounts(canonical, v)
	if err != nil {
		return nil, err
	}
	if indent {
		return json.MarshalIndent(canonical, "", "  ")
	}
	return json.Marshal(canonical)
}

// restoreUnresolvedTopLevelMounts writes an UNRESOLVED top-level `models.<k>`
// `{ref}` MOUNT EDGE back out as the edge it was authored as.
//
// `ESMFile.Models` is a `map[string]Model`, so a bare `{ref}` entry decodes to
// an EMPTY Model. LoadString snapshots the edges off the text before that
// decode and the ref resolver consumes them — but LoadString and LoadDocument
// do NOT resolve refs (only LoadPath does), so a document that came in through
// either and goes out again would otherwise emit `{"variables": null}` where
// the mount was: the `ref`, its `bindings` and its `index_set_rename` gone,
// with no error anywhere.
//
// The `subsystems.<k>` form never had this problem, because that map is
// `map[string]any` and keeps the edge verbatim. esm-spec §4.7 "Two mount forms,
// one mechanism" is what says the two must not differ here either.
//
// Only edges the resolver has NOT consumed are still in the map, so this can
// never overwrite a mount that did resolve.
func restoreUnresolvedTopLevelMounts(canonical any, v any) (any, error) {
	file, ok := v.(*ESMFile)
	if !ok || len(file.topLevelModelRefs) == 0 {
		return canonical, nil
	}
	root, ok := canonical.(map[string]any)
	if !ok {
		return canonical, nil
	}
	models, ok := root["models"].(map[string]any)
	if !ok {
		return canonical, nil
	}
	for _, name := range sortedKeys(file.topLevelModelRefs) {
		if _, present := models[name]; !present {
			continue
		}
		edge, err := canonicalizeForJSON(file.topLevelModelRefs[name])
		if err != nil {
			return nil, err
		}
		models[name] = edge
	}
	return canonical, nil
}

// serializeDocument is the shared serialization core for the four exported
// entry points (ToJSON/ToJSONCompact return the string; WritePath/
// WritePathCompact persist it). It validates the file (unlike the raw
// (*ESMFile).ToJSON METHOD, which is a plain marshal) and emits canonical
// JSON, indented when indent is true and compact otherwise.
func serializeDocument(file *ESMFile, indent bool) (string, error) {
	if file == nil {
		return "", fmt.Errorf("cannot serialize nil ESM file")
	}

	// Validate the file before serializing
	if err := file.ValidateStruct(); err != nil {
		return "", fmt.Errorf("validation failed before serialization: %w", err)
	}

	jsonData, err := marshalCanonical(file, indent)
	if err != nil {
		return "", fmt.Errorf("failed to marshal ESM file to JSON: %w", err)
	}

	return string(jsonData), nil
}

// ToJSON validates an ESM file and returns it as an indented canonical JSON
// string. PURE — it never touches disk; WritePath is the writer.
//
// Go was the only binding whose NAMES already distinguished serializing from
// writing (Serialize/SaveToFile) — and the only one whose names matched
// nobody else's. These are those functions under the shared names.
func ToJSON(file *ESMFile) (string, error) {
	out, err := serializeDocument(file, true)
	if err != nil {
		return out, err
	}
	return raiseFaqEsmFloor(out), nil
}

// raiseFaqEsmFloor raises an emitted document's declared `esm` to 1.1.0 when it
// CONTAINS a `faq` node but declares less.
//
// Like the §9.6.4 rule-8 template stamp this is a FLOOR — it only ever raises.
// A document can come to CONTAIN `faq` without ever spelling it: a 1.0.0 parent that mounts a subsystem whose child uses `faq` has the child inlined at load, and emitting that as 1.0.0 writes a document the `faq_version_too_old` gate then refuses to read back.
// The load-time gate cannot catch it — the parent's AUTHORED bytes are legal — so the stamp closes it here.
// See docs/content/rfcs/faq-node-rename.md §5.5.
func raiseFaqEsmFloor(jsonStr string) string {
	_, _, _, hasFaq := scanOpAliases(jsonStr)
	if !hasFaq {
		return jsonStr
	}
	declared, below := declaredEsmBelowV11(jsonStr)
	if !below {
		return jsonStr
	}
	return raiseEsmFloorToV11(jsonStr, declared)
}

// ToJSONCompact validates an ESM file and returns it as a compact canonical
// JSON string (no indentation). A separate function rather than a ToJSON
// option because Go has no default arguments.
func ToJSONCompact(file *ESMFile) (string, error) {
	out, err := serializeDocument(file, false)
	if err != nil {
		return out, err
	}
	return raiseFaqEsmFloor(out), nil
}

// WritePath writes an ESM file to path as indented canonical JSON. It returns
// only an error, never the payload: no function in this API both writes and
// hands back the serialized bytes — call ToJSON when you want the string.
func WritePath(file *ESMFile, path string) error {
	jsonStr, err := ToJSON(file)
	if err != nil {
		return err
	}

	// Write to file
	if err := writeFile(path, []byte(jsonStr)); err != nil {
		return fmt.Errorf("failed to write file %s: %w", path, err)
	}

	return nil
}

// WritePathCompact writes an ESM file to path in the compact canonical form.
func WritePathCompact(file *ESMFile, path string) error {
	jsonStr, err := ToJSONCompact(file)
	if err != nil {
		return err
	}

	// Write to file
	if err := writeFile(path, []byte(jsonStr)); err != nil {
		return fmt.Errorf("failed to write file %s: %w", path, err)
	}

	return nil
}

// writeFile is a simple file writing helper.
func writeFile(path string, data []byte) error {
	return os.WriteFile(path, data, 0644)
}

// SerializeExpression serializes just an expression to JSON
func SerializeExpression(expr Expression) (string, error) {
	jsonData, err := marshalCanonical(expr, true)
	if err != nil {
		return "", fmt.Errorf("failed to serialize expression: %w", err)
	}
	return string(jsonData), nil
}

// SerializeExpressionCompact serializes just an expression to compact JSON
func SerializeExpressionCompact(expr Expression) (string, error) {
	jsonData, err := marshalCanonical(expr, false)
	if err != nil {
		return "", fmt.Errorf("failed to serialize expression: %w", err)
	}
	return string(jsonData), nil
}

// SerializeModel serializes just a model to JSON
func SerializeModel(model *Model) (string, error) {
	if model == nil {
		return "", fmt.Errorf("cannot serialize nil model")
	}

	jsonData, err := marshalCanonical(model, true)
	if err != nil {
		return "", fmt.Errorf("failed to serialize model: %w", err)
	}
	return string(jsonData), nil
}

// SerializeReactionSystem serializes just a reaction system to JSON
func SerializeReactionSystem(system *ReactionSystem) (string, error) {
	if system == nil {
		return "", fmt.Errorf("cannot serialize nil reaction system")
	}

	jsonData, err := marshalCanonical(system, true)
	if err != nil {
		return "", fmt.Errorf("failed to serialize reaction system: %w", err)
	}
	return string(jsonData), nil
}

File: pkg/earthsci-ast-rs/src/serialize.rs
//! JSON serialization for ESM files

use crate::{EsmFile, error::EsmError};
use std::path::Path;

/// Serialize an ESM file to JSON string
///
/// This function converts an `EsmFile` struct back to a JSON string.
/// The output will be pretty-printed for human readability.
///
/// # Arguments
///
/// * `esm_file` - The ESM file to serialize
///
/// # Returns
///
/// * `Ok(String)` - Successfully serialized JSON string
/// * `Err(EsmError)` - Serialization error
///
/// # Examples
///
/// ```rust
/// use earthsci_ast::{EsmFile, Metadata, to_json};
///
/// let esm_file = EsmFile {
///     esm: "1.0.0".to_string(),
///     metadata: Metadata {
///         name: Some("test_model".to_string()),
///         ..Default::default()
///     },
///     ..Default::default()
/// };
///
/// let json = to_json(&esm_file).expect("Failed to serialize ESM file");
/// assert!(json.contains("\"esm\": \"1.0.0\""));
/// ```
pub fn to_json(esm_file: &EsmFile) -> Result<String, EsmError> {
    let mut value = serde_json::to_value(esm_file).map_err(EsmError::JsonParse)?;
    raise_faq_esm_floor(&mut value);
    serde_json::to_string_pretty(&value).map_err(EsmError::JsonParse)
}

/// Raise an emitted document's declared `esm` to 1.1.0 when it CONTAINS a `faq`
/// node but declares less.
///
/// Like the §9.6.4 rule-8 template stamp, this is a FLOOR — "a consumer needs at
/// least this" — so it only ever raises. It exists because a document can come
/// to contain `faq` without ever spelling it: a 1.0.0 parent that mounts a
/// subsystem whose child uses `faq` has the child inlined into it at load, and
/// emitting that as 1.0.0 writes a document the `faq_version_too_old` gate then
/// refuses to read back. The load-time gate cannot catch it — the parent's
/// AUTHORED bytes are legal — so the stamp closes it here.
/// See docs/content/rfcs/faq-node-rename.md §5.5.
fn raise_faq_esm_floor(value: &mut serde_json::Value) {
    fn has_faq(v: &serde_json::Value) -> bool {
        match v {
            serde_json::Value::Object(map) => {
                map.get("op").and_then(|o| o.as_str()) == Some("faq")
                    || map.values().any(has_faq)
            }
            serde_json::Value::Array(items) => items.iter().any(has_faq),
            _ => false,
        }
    }
    let below = value
        .get("esm")
        .and_then(|v| v.as_str())
        .and_then(crate::diagnostic::parse_semver)
        .is_some_and(|(major, minor, _)| (major, minor) < (1, 1));
    if below && has_faq(value)
        && let Some(obj) = value.as_object_mut()
    {
        obj.insert("esm".to_string(), serde_json::Value::String("1.1.0".to_string()));
    }
}

/// Serialize an ESM file to compact JSON string (no pretty printing)
///
/// This function is similar to [`to_json`] but produces compact JSON without
/// extra whitespace, suitable for storage or transmission. It is a separate
/// function rather than a `to_json(file, opts)` flag because Rust has no
/// default arguments and a one-field options struct plus a
/// `to_json_with_options` twin is heavier than the pair.
///
/// # Arguments
///
/// * `esm_file` - The ESM file to serialize
///
/// # Returns
///
/// * `Ok(String)` - Successfully serialized compact JSON string
/// * `Err(EsmError)` - Serialization error
pub fn to_json_compact(esm_file: &EsmFile) -> Result<String, EsmError> {
    let mut value = serde_json::to_value(esm_file).map_err(EsmError::JsonParse)?;
    raise_faq_esm_floor(&mut value);
    serde_json::to_string(&value).map_err(EsmError::JsonParse)
}

/// Write an ESM file to `path` as pretty-printed JSON.
///
/// Returns `Ok(())`, never the payload: no function in this API both writes
/// and hands back the serialized bytes — call [`to_json`] when you want the
/// string. (`save` used to return the string here and in TypeScript while
/// WRITING TO DISK in Julia, under the same name.)
pub fn write_path<P: AsRef<Path>>(esm_file: &EsmFile, path: P) -> Result<(), EsmError> {
    let path = path.as_ref();
    let json = to_json(esm_file)?;
    std::fs::write(path, json).map_err(|e| EsmError::FileWrite {
        path: path.display().to_string(),
        source: e,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::var;
    use crate::types::{Equation, Metadata, ModelVariable, VariableType};
    use crate::{Expr, Model};
    use indexmap::IndexMap;

    #[test]
    fn test_save_minimal_file() {
        let esm_file = EsmFile {
            esm: "0.1.0".to_string(),
            metadata: Metadata {
                name: Some("test_model".to_string()),
                ..Default::default()
            },
            ..Default::default()
        };

        let result = to_json(&esm_file);
        assert!(result.is_ok());

        let json = result.unwrap();
        assert!(json.contains("\"esm\": \"0.1.0\""));
        assert!(json.contains("\"name\": \"test_model\""));
    }

    #[test]
    fn test_save_with_model() {
        let mut models = IndexMap::new();
        let mut variables = IndexMap::new();
        variables.insert(
            "x".to_string(),
            ModelVariable {
                default: Some(0.0.into()),
                ..var(VariableType::Unknown, Some("m"))
            },
        );

        models.insert(
            "test".to_string(),
            Model {
                name: Some("Test Model".to_string()),
                variables,
                equations: vec![Equation {
                    comment: None,
                    lhs: Expr::Variable("d(x)/dt".to_string()),
                    rhs: Expr::Number(1.0),
                }],
                ..Default::default()
            },
        );

        let esm_file = EsmFile {
            esm: "0.1.0".to_string(),
            metadata: Metadata {
                name: Some("test_model".to_string()),
                ..Default::default()
            },
            models: Some(models),
            ..Default::default()
        };

        let result = to_json(&esm_file);
        assert!(result.is_ok());

        let json = result.unwrap();
        assert!(json.contains("\"models\""));
        assert!(json.contains("\"test\""));
        assert!(json.contains("\"variables\""));
        assert!(json.contains("\"equations\""));
    }

    #[test]
    fn test_save_compact() {
        let esm_file = EsmFile {
            esm: "0.1.0".to_string(),
            metadata: Metadata {
                name: Some("test_model".to_string()),
                ..Default::default()
            },
            ..Default::default()
        };

        let result = to_json_compact(&esm_file);
        assert!(result.is_ok());

        let json = result.unwrap();
        // Compact JSON shouldn't have extra whitespace
        assert!(!json.contains("  "));
        assert!(json.contains("\"esm\":\"0.1.0\""));
    }
}
